### PR TITLE
Add custom wield item animation using linear keyframes (WIP)

### PR DESCRIPTION
**Pull request by rubenwardy**
``` lua
minetest.register_node("default:apple", {
    --
    wield_keyframes = {
        {
            x = 0,
            y = 0,
            z = 20,
            duration = 5
        }
    },
    --
})
```

Todo:
- rotation in key frames
- remove logging
- more than linear (bezier curves / splines)
- particles etc
- documentation

### CODE REVIEW

**Comment by kaadmy** [comment]
:+1:

**Comment by benrob0329** [comment]
Can you specify which frame to show? (Eg for a compass)

**Comment by rubenwardy** [comment]
No, this is only for dig/punch animations

**Comment by bendeutsch** [comment]
I've currently got a similar enhancement [in the works](https://github.com/bendeutsch/minetest/tree/wield-animations), except using splines instead of keyframes. Maybe these systems could be combined?

**Comment by rubenwardy** [comment]
Splines would be even better - this is mostly a proof of concept. Feel free to nick stuff from here if it's useful

**Comment by bendeutsch** [comment]
It should be relatively easy to have both – keyframes are basically a sequence of splines of degree 1. I don't have any lua bindings or netcode, so either I'll nick those, or (even better) I'll send you a pull request for my spline stuff ;-)

One thing I'm planning, though, is a set of "basic" animations (dig up, dig down, eat, poke etc.), so that we don't need every food item to copy-and-paste the same eating keyframes. I can also make a pull request into your pull request, if you want.

**Comment by rubenwardy** [comment]
Sure! That would be greate

**Comment by Ferk** [comment]
For what I understand, this would replace the movement that the wield_item does when "on_use" is triggered. This would be awesome for custom eating movement, reading books, etc.

I haven't been able to test it, but some questions:

* would it also affect the animation done when pressing right click?

If so, this might be a bit annoying when you want to open your chest with an apple or a book in your hands, since the animation makes you think you are eating or blocks your view for a second putting the book in front of you.

* would the duration of the animation also add a delay to the action taking place? and if so, would it be possible to cancel the animation mid-way and thus the action if the mouse is released early?

I would really like it if this was the case. Since a slow eating/reading animation that can be cancelled would prevent a lot of misclicks.
Also, this could potentially add some variability to how different weapons are handled, maybe improving a bit the combat mechanics.

**Comment by 0-afflatus** [comment]
Nice idea even in this limited form.

**Comment by bendeutsch** [comment]
Ferk, I agree that it is *very* desirable to add such a delay. Ideally, this should be possible client-side – the server need never know that the client waited a bit before sending the "use" action. Or not, if the user released the button beforehand.

Good point about the different mouse buttons – how many actions would one have to differentiate? Left click, right click, the use button, and shifted versions of all of them? I would like to be able to explicitly override an item's digging animation; the shovel's downward swing has always annoyed me.

**Comment by DonBatman** [comment]
This would be great for many things. 
Daggers jab
Screwdriver goes straight and maybe turns
Guns stay straight but has recoil
Etc

File: src/camera.cpp [hunk 1]
```diff
@@ -407,8 +424,7 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 busytime,
 	//v3f wield_rotation = v3f(-100, 120, -100);
 	v3f wield_rotation = v3f(-100, 120, -100);
 	wield_position.Y += fabs(m_wield_change_timer)*320 - 40;
-	if(m_digging_anim < 0.05 || m_digging_anim > 0.5)
-	{
+	if(m_digging_anim < 0.05 || m_digging_anim > 0.5) {
```
Comment (bendeutsch) [inline src/camera.cpp]: Incidently, does anyone know what this block actually _does_?

Comment (rubenwardy) [inline src/camera.cpp]: I honestly have no idea. I forgot about this bit

Comment (rubenwardy) [inline src/camera.cpp]: Apparently it's a tool recharge animation. https://github.com/minetest/minetest/commit/989aba1966373885d4cac306daea25e102f8d68d

@celeron55, please may you explain?

File: src/wieldmesh.h [hunk 2]
```diff
@@ -44,6 +44,8 @@ class WieldMeshSceneNode: public scene::ISceneNode
 			v3f wield_scale, ITextureSource *tsrc, u8 num_frames);
 	void setItem(const ItemStack &item, IGameDef *gamedef);
 
+	ItemStack getItem();
```
Comment (bendeutsch) [inline src/wieldmesh.h]: I can't seem to find the definition for this…?

Comment (rubenwardy) [inline src/wieldmesh.h]: that was a mistake, didn't mean to add that (I was going to get the keyframes another way, but then realised that there was no need)

Comment (bendeutsch) [inline src/wieldmesh.h]: Oh, I don't know – I would like a way to access the _current_ wield item at any time, not just the _next_ one (or am I missing something?)

**Comment by cdqwertz** [comment]
:+1:

**Comment by sfan5** [comment]
This should probably be rewritten to make use of the `TileAnimation` code.

**Comment by Ferk** [comment]
For what I understand, the ``TileAnimation`` code concerns animation using spritesheet, changing textures.

I think this is about the movement in the hand of the wielditem, it doesn't really animate the textures or the wielditem itself, it just changes its movement. I'm not sure if adding TileAnimation-based animation would be in the scope of this PR.

**Comment by rubenwardy** [comment]
Closing for now, as out of date and needs more work

**Comment by sorcerykid** [comment]
It looks like it might be possible to revive @bendeutsch 's excellent spline sequence class for a PR related to #9717. That is some excellent work :)

**Comment by bendeutsch** [comment]
Yes, by all means take the spline classes, please! They were meant to be general-purpose utilities. I just don't have the time to do anything about it myself, sorry.

**Pull request closed without merging**